README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# svgfilter <img src="man/figures/logo.svg" align="right" height=230/>

<!-- badges: start -->

![](http://img.shields.io/badge/cool-useless-green.svg)
![](http://img.shields.io/badge/devout-verse-blue.svg)
<!-- badges: end -->

`svgfilter` provides some basic SVG filters.

Because github sanitises SVG to remove some elements of style, scripting
and animation, please see the [svgfilter pkgdown
website](https://coolbutuseless.github.io/package/svgfilter/index.html)
to view the animations.

Creating interesting filters is a bit of an art-form, so this package
provides a few simple filters to show how they operate.

This packages is not at all meant to be a comprehensive set of all
possible filters, but merely some examples to show the way.

The following filters are provided in this package:

  - Turbulent displacement (as shown in the hex logo) -
    `create_filter_turbulent_displacement()`
  - Gaussian Blur - `create_filter_blur()`
  - Drop Shadow - `create_filter_drop_shadow()`

## Installation

You can install from
[GitHub](https://github.com/coolbutuseless/svgfilter) with:

``` r
# install.packages("devtools")
devtools::install_github("coolbutuseless/minisvg")   # Build SVG documents with R
devtools::install_github("coolbutuseless/svgfilter") # This package
```

## Blur filter

``` r
my_filter <- create_filter_blur(id = "example-blur")

my_filter$save_full_svg("man/figures/README-eg-blur.svg", height=100, include_declaration = FALSE)
```

<pre><details closed><summary style='color: #4169E1;'> Show/hide SVG text </summary>&lt;filter id="example-blur" x="-20%" y="-20%" width="140%" height="150%"&gt;
  &lt;feGaussianBlur in="SourceGraphic" stdDeviation="3" /&gt;
&lt;/filter&gt;</details></pre>

<img src = "man/figures/README-eg-blur.svg">

#### Applying the filter

``` r
doc <- svg_doc(width=300, height = 200)
doc$defs(my_filter)

doc$circle(
  id     = "my circle",
  cx     = 100, 
  cy     = 100,
  r      = 50, 
  fill   = '#002366',
  filter = my_filter
)

doc$save("man/figures/README-doc-blur.svg")
```

<pre><details closed><summary style='color: #4169E1;'> Show/hide SVG text </summary>&lt;?xml version="1.0" encoding="UTF-8"?&gt;
&lt;svg width="300" height="200" viewBox="0 0 300 200" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink"&gt;
  &lt;defs&gt;
    &lt;filter id="example-blur" x="-20%" y="-20%" width="140%" height="150%"&gt;
      &lt;feGaussianBlur in="SourceGraphic" stdDeviation="3" /&gt;
    &lt;/filter&gt;
  &lt;/defs&gt;
  &lt;circle id="my circle" fill="#002366" filter="url('#example-blur')" cx="100" cy="100" r="50" /&gt;
&lt;/svg&gt;</details></pre>

<img src = "man/figures/README-doc-blur.svg">

## Drop Shadow filter

``` r
my_filter <- create_filter_drop_shadow(id = "example-drop")

my_filter$save_full_svg("man/figures/README-eg-drop.svg", height=100, include_declaration = FALSE)
```

<pre><details closed><summary style='color: #4169E1;'> Show/hide SVG text </summary>&lt;filter id="example-drop" x="-20%" y="-20%" width="140%" height="150%"&gt;
  &lt;feOffset in="SourceAlpha" dx="10" dy="10" result="offset_output" /&gt;
  &lt;feGaussianBlur in="offset_output" stdDeviation="3" result="blur_output" /&gt;
  &lt;feFlood flood-color="#bbb" result="flood_image" /&gt;
  &lt;feComposite in="flood_image" in2="blur_output" operator="in" result="composite_output" /&gt;
  &lt;feBlend in="SourceGraphic" in2="composite_output" mode="normal" /&gt;
&lt;/filter&gt;</details></pre>

<img src = "man/figures/README-eg-drop.svg">

#### Applying the filter

``` r
doc <- svg_doc(width=300, height = 200)
doc$defs(my_filter)

doc$circle(
  id     = "my circle",
  cx     = 100, 
  cy     = 100,
  r      = 50, 
  fill   = '#002366',
  filter = my_filter
)

doc$save("man/figures/README-doc-drop.svg")
```

<pre><details closed><summary style='color: #4169E1;'> Show/hide SVG text </summary>&lt;?xml version="1.0" encoding="UTF-8"?&gt;
&lt;svg width="300" height="200" viewBox="0 0 300 200" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink"&gt;
  &lt;defs&gt;
    &lt;filter id="example-drop" x="-20%" y="-20%" width="140%" height="150%"&gt;
      &lt;feOffset in="SourceAlpha" dx="10" dy="10" result="offset_output" /&gt;
      &lt;feGaussianBlur in="offset_output" stdDeviation="3" result="blur_output" /&gt;
      &lt;feFlood flood-color="#bbb" result="flood_image" /&gt;
      &lt;feComposite in="flood_image" in2="blur_output" operator="in" result="composite_output" /&gt;
      &lt;feBlend in="SourceGraphic" in2="composite_output" mode="normal" /&gt;
    &lt;/filter&gt;
  &lt;/defs&gt;
  &lt;circle id="my circle" fill="#002366" filter="url('#example-drop')" cx="100" cy="100" r="50" /&gt;
&lt;/svg&gt;</details></pre>

<img src = "man/figures/README-doc-drop.svg">

## Turbulent Displacement Filter (with animation)

This is quite a complex filter which is made up of the following
elements:

  - create some random turbulence
  - generate a number of different versions of the turbulence by
    changing the random seed
  - use the turbulence to displace the element the filter is applied to.

The user can specify the frequency of the noise in the x and y
directions and other noise parameters, such as the scale factor used for
the displacement.

``` r
my_filter <- create_filter_turbulent_displacement(
  id    = "example-turbulence", 
  freqx = 0.10, 
  freqy = 0.01,
  scale = 20
)
```

<pre><details closed><summary style='color: #4169E1;'> Show/hide SVG text </summary>&lt;filter id="example-turbulence" x="-30%" y="-30%" width="160%" height="160%"&gt;
  &lt;feTurbulence type="turbulence" baseFrequency="0.1 0.01" numOctaves="4" seed="1" result="turbulence"&gt;
    &lt;animate attributeName="seed" dur="2.5" repeatCount="indefinite" calcMode="discrete" values="1;2;3;4;5;6;7;8;9;10;11;12;13;14;15;16;17;18;19;20;1" keyTimes="0;0.05;0.1;0.15;0.2;0.25;0.3;0.35;0.4;0.45;0.5;0.55;0.6;0.65;0.7;0.75;0.8;0.85;0.9;0.95;1" /&gt;
  &lt;/feTurbulence&gt;
  &lt;feDisplacementMap in="SourceGraphic" in2="turbulence" scale="20" xChannelSelector="R" yChannelSelector="G" /&gt;
&lt;/filter&gt;</details></pre>

``` r
my_filter$show()
```

<img src = "man/figures/README-eg-turbulence.svg">

#### Applying the filter

``` r
doc <- svg_doc(width=300, height = 200)
doc$defs(my_filter)

doc$circle(
  id     = "my circle",
  cx     = 100, 
  cy     = 100,
  r      = 50, 
  fill   = '#002366',
  filter = my_filter
)

doc$save("man/figures/README-doc-turbulence.svg")
```

<pre><details closed><summary style='color: #4169E1;'> Show/hide SVG text </summary>&lt;?xml version="1.0" encoding="UTF-8"?&gt;
&lt;svg width="300" height="200" viewBox="0 0 300 200" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink"&gt;
  &lt;defs&gt;
    &lt;filter id="example-turbulence" x="-30%" y="-30%" width="160%" height="160%"&gt;
      &lt;feTurbulence type="turbulence" baseFrequency="0.1 0.01" numOctaves="4" seed="1" result="turbulence"&gt;
        &lt;animate attributeName="seed" dur="2.5" repeatCount="indefinite" calcMode="discrete" values="1;2;3;4;5;6;7;8;9;10;11;12;13;14;15;16;17;18;19;20;1" keyTimes="0;0.05;0.1;0.15;0.2;0.25;0.3;0.35;0.4;0.45;0.5;0.55;0.6;0.65;0.7;0.75;0.8;0.85;0.9;0.95;1" /&gt;
      &lt;/feTurbulence&gt;
      &lt;feDisplacementMap in="SourceGraphic" in2="turbulence" scale="20" xChannelSelector="R" yChannelSelector="G" /&gt;
    &lt;/filter&gt;
  &lt;/defs&gt;
  &lt;circle id="my circle" fill="#002366" filter="url('#example-turbulence')" cx="100" cy="100" r="50" /&gt;
&lt;/svg&gt;</details></pre>

<img src = "man/figures/README-doc-turbulence.svg">

## Creating the logo for this package

``` r
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Building an SVG logo with an animated stripe
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
logo <- svg_doc(width = 220, height = 220)

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Create a rotating striped pattern and add it to the '<defs>' 
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
turbulent_displacement_filter <- svgfilter::create_filter_turbulent_displacement(
  id = "tdfilter",
  scale = 20
)

gradient_pattern <- svgpatternsimple::create_pattern_gradient(
  id = "red-gradient",
  colour1 = 'red',
  colour2 = 'gold',
  angle   = 90
)

logo$defs(gradient_pattern)
logo$defs(turbulent_displacement_filter)

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Create a hexagon filled with this pattern, and add it to the document
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
len     <- 100
angles  <- (seq(0, 360, 60) + 90) * pi/180
xs      <- round(len * cos(angles) + 110, 2)
ys      <- round(len * sin(angles) + 110, 2)
hex     <- stag$polygon(id = 'hex', xs = xs, ys = ys)

hex$update(stroke = '#c0c0c0', filter = turbulent_displacement_filter,
           fill = gradient_pattern)
# hex$update(stroke = 'black', fill_opacity = 0)

logo$append(hex)

# logo$show()
logo$save("man/figures/logo.svg")
```

<pre><details closed><summary style='color: #4169E1;'> Show/hide SVG text </summary>&lt;?xml version="1.0" encoding="UTF-8"?&gt;
&lt;svg width="220" height="220" viewBox="0 0 220 220" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink"&gt;
  &lt;defs&gt;
    &lt;linearGradient id="red-gradient" x1="0%" y1="100%" x2="0%" y2="0%"&gt;
      &lt;stop offset="0%" style="stop-color:red;stop-opacity:1" /&gt;
      &lt;stop offset="100%" style="stop-color:gold;stop-opacity:1" /&gt;
    &lt;/linearGradient&gt;
  &lt;/defs&gt;
  &lt;defs&gt;
    &lt;filter id="tdfilter" x="-30%" y="-30%" width="160%" height="160%"&gt;
      &lt;feTurbulence type="turbulence" baseFrequency="0.05" numOctaves="4" seed="1" result="turbulence"&gt;
        &lt;animate attributeName="seed" dur="2.5" repeatCount="indefinite" calcMode="discrete" values="1;2;3;4;5;6;7;8;9;10;11;12;13;14;15;16;17;18;19;20;1" keyTimes="0;0.05;0.1;0.15;0.2;0.25;0.3;0.35;0.4;0.45;0.5;0.55;0.6;0.65;0.7;0.75;0.8;0.85;0.9;0.95;1" /&gt;
      &lt;/feTurbulence&gt;
      &lt;feDisplacementMap in="SourceGraphic" in2="turbulence" scale="20" xChannelSelector="R" yChannelSelector="G" /&gt;
    &lt;/filter&gt;
  &lt;/defs&gt;
  &lt;polygon points="110,210 23.4,160 23.4,60 110,10 196.6,60 196.6,160 110,210" id="hex" stroke="#c0c0c0" filter="url('#tdfilter')" fill="url('#red-gradient')" /&gt;
&lt;/svg&gt;</details></pre>

<img src = "man/figures/logo.svg">


NEWS.md
# svgfilter 0.1.0  2020-01-26

* Initial release


working/test.R


library(minisvg)


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Define filter with turbulence driving the displacmenet
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
my_filter <- stag$filter(
  id = "displacementFilter",
  stag$feTurbulence(
    type          = "turbulence",
    baseFrequency = 0.05,
    numOctaves    = 2,
    result        = "turbulence"),
  stag$feDisplacementMap(
    in_   = "SourceGraphic",
    in2   = "turbulence",
    scale = 50,
    xChannelSelector = 'R',
    yChannelSelector = 'G',
    stag$animate(attributeName = 'scale', values="20;70;40;50;20", dur="3s", repeatCount = "indefinite"),
    NULL
  )
)


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Apply this displacement filter to a circle
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
doc <- svg_doc(
  width = 400, height = 400,
  my_filter,
  stag$circle(cx=200, cy=200, r=80, filter = my_filter,
              stag$animateTransform(attributeName = "transform", attributeType = 'XML',
                                    type = "rotate", from = "0 200 200", to = "360 200 200 ",
                                    dur = "7s", repeatCount = "indefinite"))
)


doc$show()


R/filter-blur.R


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Create an SVG blur filter
#'
#' Create a filter to apply Gaussian blur to an element.
#'
#' For more information see the MDN SVG docs for
#' \href{https://developer.mozilla.org/en-US/docs/Web/SVG/Element/feGaussianBlur}{feGaussianBlur}
#'
#'
#' @param id id to use for filter.
#' @param std_dev standard deviation of the gaussian blue. default: 3
#' @param ... other arguments ignored
#'
#' @return minisvg::SVGElement object representing a filter
#'
#' @import minisvg
#' @export
#'
#'
#' @examples
#' \dontrun{
#' # Create an SVG document
#' library(minisvg)
#' doc   <- minisvg::svg_doc()
#'
#' # Create the filter and add to the SVG definitions
#' blur_filter <- create_filter_blur(id = 'myblur')
#' doc$defs(blur_filter)
#'
#' # Create a rectangle with the animation
#' rect  <- stag$rect(
#'   x      = "10%",
#'   y      = "10%",
#'   width  = "80%",
#'   height = "80%",
#'   fill   = "lightblue",
#'   stroke = 'black',
#'   filter = blur_filter
#' )
#'
#' # Add this rectangle to the document, show the SVG text, then render it
#' doc$append(rect)
#' doc
#' doc$show()
#' }
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
create_filter_blur <- function(id, std_dev = 3, ...) {

  blur <- minisvg::SVGFilter$new(
    id = id,
    x = "-20%", y = "-20%", width="140%", height="150%",
    stag$feGaussianBlur(
      in_          = 'SourceGraphic',
      stdDeviation = std_dev
    )
  )

  blur
}



if (FALSE) {
  pat <- create_filter_blur(id = 'hello')
  pat
  pat$show(width=400, height = 200)
  pat$save_full_svg("working/one.svg", width=400, height = 100)
}


working/fire01.R


library(minisvg)

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Define filter with turbulence driving the displacmenet
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
my_filter <- stag$filter(
  id = "displacementFilter",
  x = "-30%", y = "-30%", width="160%", height="160%",
  stag$feTurbulence(
    type          = "turbulence",
    baseFrequency = 0.05,
    numOctaves    = 5,
    seed          = 2,
    stag$animate(attributeName = 'seed', values=1:20, dur="2s", keyTimes = seq(0, 1, length.out = 20), repeatCount = "indefinite"),
    result        = "turbulence"),
  stag$feDisplacementMap(
    in_   = "SourceGraphic",
    in2   = "turbulence",
    scale = 50,
    yChannelSelector = 'G',
    # stag$animate(attributeName = 'scale', values="0;200;0", dur="10s", repeatCount = "indefinite"),
    NULL
  )
)

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Red to gold vertical linear gradient
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
fire_grad <- stag$linearGradient(
  id = "myGradient",
  gradientTransform = "rotate(90)",
  stag$stop(offset =  "5%", stop_color = "gold"),
  stag$stop(offset = "95%", stop_color = "red" )
)

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Apply this displacement filter to a small rect with a red gradient fill
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
doc <- svg_doc(
  width = 400, height = 400,
  my_filter,
  stag$defs(
    my_filter,
    fire_grad
  ),
  stag$rect(
    x      = 100,
    y      = 100,
    width  = 200,
    height = 20,
    fill   = fire_grad,
    filter = my_filter,
    NULL
  )
)

doc$show()




working/fire03.R


library(minisvg)

fire_w <- 400
fire_h <- 400
scale  <- fire_h/2
dx     <- 0.5 * scale



#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Define filter with turbulence driving the displacmenet
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
my_filter <- stag$filter(
  id = "displacementFilter",
  x = "-30%", y = "-30%", width="160%", height="160%",
  stag$feTurbulence(
    type          = "turbulence",
    baseFrequency = 0.05,
    numOctaves    = 5,
    seed          = 2,
    stag$animate(attributeName = 'seed', values=1:20, dur="2s", keyTimes = seq(0, 1, length.out = 20), repeatCount = "indefinite"),
    result        = "turbulence"),
  stag$feDisplacementMap(
    in_   = "SourceGraphic",
    in2   = "turbulence",
    scale = scale,
    xChannelSelector = 'R',
    yChannelSelector = 'G',
    # stag$animate(attributeName = 'scale', values="0;200;0", dur="10s", repeatCount = "indefinite"),
    NULL
  )
)

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Red to gold vertical linear gradient
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
fire_grad <- stag$linearGradient(
  id = "myGradient",
  gradientTransform = "rotate(90)",
  stag$stop(offset =  "5%", stop_color = "gold"),
  stag$stop(offset = "95%", stop_color = "red" )
)


fire_rect <- stag$rect(
  x      = dx - dx,
  y      = dx,
  width  = fire_w + 2 * dx,
  height = fire_h,
  fill   = fire_grad,
  filter = my_filter,
  NULL
)


debug_rect <- stag$rect(
  x      = dx,
  y      = dx,
  width  = fire_w,
  height = fire_h - dx,
  stroke = 'black',
  fill_opacity = 0,
  NULL
)


extent_rect <- stag$rect(
  x = 0, y = 0,
  width = "100%",
  height = "100%",
  fill_opacity = 0,
  stroke = 'red'
)

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Apply this displacement filter to a small rect with a red gradient fill
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
doc <- svg_doc(
  width = fire_w + 2*dx, height = fire_h + 2*dx,
  stag$defs(
    my_filter,
    fire_grad
  ),
  fire_rect,
  debug_rect,
  extent_rect
)

doc$show()




R/filter-drop-shadow.R


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Create an SVG drop shadow filter
#'
#' This filter adds a drop shadow to the element it is applied to.
#'
#' The drop shadow effect is a combination of 4 primitive SVG filters:
#'
#' \enumerate{
#' \item{Blur the element}
#' \item{Offset this blurred image}
#' \item{Fill the blurred image with the given colour using a flood and a composite}
#' \item{Layer the drop shadow behind the original image}
#' }
#'
#'
#' For more information see the MDN SVG docs for
#' \href{https://developer.mozilla.org/en-US/docs/Web/SVG/Element/feGaussianBlur}{feGaussianBlur}
#'
#'
#' @param id id to use for filter.
#' @param dx,dy offset of shadow in x and y directions. default: 10
#' @param std_dev standard deviation of the gaussian for the blur. default: 3
#' @param colour default: #bbb
#' @param ... other arguments ignored
#'
#' @return minisvg::SVGElement object representing a filter
#'
#' @import minisvg
#' @export
#'
#'
#' @examples
#' \dontrun{
#' # Create an SVG document
#' library(minisvg)
#' doc   <- minisvg::svg_doc()
#'
#' # Create the filter and add to the SVG definitions
#' drop_shadow_filter <- create_filter_drop_shadow(id = 'dshadow')
#' doc$defs(drop_shadow_filter)
#'
#' # Create a rectangle with the animation
#' rect  <- stag$rect(
#'   x      = "10%",
#'   y      = "10%",
#'   width  = "80%",
#'   height = "80%",
#'   fill   = "lightblue",
#'   stroke = 'black',
#'   filter = drop_shadow_filter
#' )
#'
#' # Add this rectangle to the document, show the SVG text, then render it
#' doc$append(rect)
#' doc
#' doc$show()
#' }
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
create_filter_drop_shadow <- function(id, dx = 10, dy = dx,
                                      std_dev = 3, colour = '#bbb', ...) {

  drop <- minisvg::SVGFilter$new(
    id = id,
    x = "-20%", y = "-20%", width="140%", height="150%",
    stag$feOffset(
      in_    = 'SourceAlpha',
      result = 'offset_output',
      dx     = dx,
      dy     = dy
    ),
    stag$feGaussianBlur(
      in_          = 'offset_output',
      result       = 'blur_output',
      stdDeviation = std_dev
    ),
    stag$feFlood(
      flood_color = colour,
      result      = 'flood_image'
    ),
    stag$feComposite(
      in_     = 'flood_image',
      in2     = 'blur_output',
      operator = 'in',
      result   = 'composite_output'
    ),
    stag$feBlend(
      in_  = 'SourceGraphic',
      in2  = 'composite_output',
      mode = 'normal'
    )
  )

  drop
}



if (FALSE) {
  pat <- create_filter_drop_shadow(id = 'hello', colour = 'red')
  pat
  pat$show(width=400, height = 200)
  # pat$save_full_svg("working/one.svg", width=400, height = 100)
}


R/filter-turbulent-displacement.R


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Create an animated SVG turbulent displacement filter
#'
#' Use turblence to displace an element.
#'
#'
#' The turbulent displacement effect is a combination of 2 primitive SVG filters:
#'
#' \enumerate{
#' \item{Create turbulence}
#' \item{Use the turblence as a displacment map for the object}
#' }
#'
#'
#' For more information see the MDN SVG docs for
#' \href{https://developer.mozilla.org/en-US/docs/Web/SVG/Element/feGaussianBlur}{feTurbulence}
#' and \href{https://developer.mozilla.org/en-US/docs/Web/SVG/Element/feDisplacementMap}{feDisplacementMap}
#'
#' @param id id to use for filter.
#' @param scale scale of the displacement
#' @param octaves corresponds to \code{feTurbulence - numOctaves} parameter. default: 4
#' @param fps frames per second. default: 8
#' @param freqx,freqy base frequency in x and y directions. default: freqx = 0.05, freqy = freqx
#' @param frames number of frames. default: 20
#' @param ... other arguments ignored
#'
#' @return minisvg::SVGElement object representing a filter
#'
#' @import minisvg
#' @export
#'
#'
#' @examples
#' \dontrun{
#' # Create an SVG document
#' library(minisvg)
#' doc   <- minisvg::svg_doc()
#'
#' # Create the filter and add to the SVG definitions
#' turb_filter <- create_filter_turbulent_displacement(id = 'turb')
#' doc$defs(turb_filter)
#'
#' # Create a rectangle with the animation
#' rect  <- stag$rect(
#'   x      = "10%",
#'   y      = "10%",
#'   width  = "80%",
#'   height = "80%",
#'   fill   = "lightblue",
#'   stroke = 'black',
#'   filter = turb_filter
#' )
#'
#' # Add this rectangle to the document, show the SVG text, then render it
#' doc$append(rect)
#' doc
#' doc$show()
#' }
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
create_filter_turbulent_displacement <- function(id,
                                                 scale   = 50,
                                                 octaves = 4,
                                                 fps     = 8,
                                                 freqx   = 0.05,
                                                 freqy   = freqx,
                                                 frames  = 20,
                                                 ...) {


  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # Red to gold vertical linear gradient
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  if (fps == 0) {
    keyTimes    <- NULL
    duration    <- NULL
    seed_values <- 1
  } else {

    keyTimes <- seq(0, 1, length.out = frames+1) # anim must go from 0 to 1
    duration <- frames/fps
    if (!is.finite(duration)) {
      stop("create_turbulent_displacment_filter(): not a finite duration given frames=", frames, ", fps=", fps,
           call. = FALSE)
    }

    seed_values <- c(seq(frames), 1) # anim must loop back to stat
  }

  freq <- paste(unique(c(freqx, freqy)), collapse = " ")

  turbulent_displacement <- minisvg::SVGFilter$new(
    id = id,
    x = "-30%", y = "-30%", width="160%", height="160%",
    stag$feTurbulence(
      type          = "turbulence",
      baseFrequency = freq,
      numOctaves    = octaves,
      seed          = 1,
      stag$animate(
        attributeName = 'seed',
        values        = seed_values,
        dur           = duration,
        keyTimes      = keyTimes,
        repeatCount   = "indefinite",
        calcMode      = 'discrete'
      ),
      result        = "turbulence"),
    stag$feDisplacementMap(
      in_   = "SourceGraphic",
      in2   = "turbulence",
      scale = scale,
      xChannelSelector = 'R',
      yChannelSelector = 'G',
      NULL
    )
  )

  turbulent_displacement
}



if (FALSE) {
  pat <- create_filter_turbulent_displacement(id = 'hello', freqx = 0.01, freqy = 0.1)
  pat
  pat$show(width=400, height = 200)
  pat$save_full_svg("working/one.svg", width=400, height = 100)
}
